drivers: net: loopback: Network loopback interface driver

pfalcon opened the pull request:
Loopback is a networking interface which doesn't actually transfer
any data via link layer externally, and instead just mirrors back
(i.e. any packet send to the loopback interface will be received from
it). This interface very useful for testing.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
drivers: net: loopback: Network loopback interface driver

Loopback is a networking interface which doesn't actually transfer
any data via link layer externally, and instead just mirrors back
(i.e. any packet send to the loopback interface will be received from
it). This interface very useful for testing.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Together with https://github.com/zephyrproject-rtos/zephyr/pull/4457, this allows to write clean self-contained (being able to run via sanitycheck) networking tests.

Based on some interface code from tests/net/.

pfalcon commented:
To use in an app:

~~~
+CONFIG_NET_TEST=y
+CONFIG_NET_LOOPBACK=y
~~~

nashif commented:
would be great to start gathering all of the net related drivers into one category, i.e. net/ and add slip and this one in there, having a category for each type, especially when you end up having just one driver for each is not nice.

pfalcon commented:
@nashif, though about that. I assume `drivers/ethernet/` and `drivers/ieee802154/` are good, and we talk about consolidating "misc network drivers". `drivers/net/`sounds good. Then specific steps:

1. Rename drivers/slip/ to drivers/net/.
2. Put new loopback driver there.

@nashif , @jukkar , @tbursztyka : Please ack.

nashif commented:
ack

pfalcon committed:
drivers: slip: Consolidate under drivers/net/

Make drivers/net/ be the place for misc networking-related drivers
(otherwise, we'd need to have a new dir per driver).

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Reworked as suggested.

There appears to be few unrelated false positives in the CI.

nashif commented:
recheck

jukkar commented on drivers/Makefile:
```diff
@@ -19,7 +19,7 @@ obj-$(CONFIG_I2S) += i2s/
 obj-$(CONFIG_ADC) += adc/
 obj-$(CONFIG_NET_L2_ETHERNET) += ethernet/
-obj-$(CONFIG_SLIP) += slip/
+obj-y += net/
```
We could also have
`obj-$(CONFIG_NETWORKING) += net/`
here.

pfalcon replied:
Done.

jukkar commented on drivers/net/loopback.c:
```diff
@@ -0,0 +1,83 @@
+	/* We need to swap the IP addresses because otherwise
+	 * the packet will be dropped.
+	 */
+
```
Swapping IP address is probably not enough for UDP / TCP data, so the port numbers should be swapped too.

pfalcon replied:
It actually made me ponder what it means to implement a loopback interface. Based on its description (everything which gets send to it, gets received from it), it must swap IP addresses. But loopback interface is not an emulator of TCP/UDP echo servers. Like, you can't do on Linux:

~~~
nc -u 127.0.0.1 56687
~~~

and expect that data we input gets back to us. So, I guess swapping just addresses should be good enough.

jukkar replied:
That is certainly true. We can leave it like this, it is just that depending on what one wants to do with this loopback interface, the application might need to know what is going on.

pfalcon replied:
Yeah, I'd suggest we start like that, and see where it leads us. An immediate usecase is just having an interface we can bind to, which doesn't have any external requirements to run (like specific hardware, a free UART, etc.)

jukkar commented on drivers/net/loopback.c:
```diff
@@ -0,0 +1,83 @@
+
+static void loopback_init(struct net_if *iface)
+{
+	net_if_set_link_addr(iface, "123456", 6, NET_LINK_DUMMY);
```
Although the mac address value does not really matter here, we could use the addresses specified for documentation. The slip driver is using these numbers so we could do things similarly.
See RFC 7042 chapter 2.1.1 https://tools.ietf.org/html/rfc7042#section-2.1.1 for details.

pfalcon replied:
So, Linux `lo` interface doesn't have any "MAC addresses". I make this call, because that's the call which puts the iface "up". And I initially tried to pass 1-byte MAC address, but anything less than 6 bytes leads to a crash in some thread later.

I'll switch to an address you refer to, but my idea was to use something simple and static.

jukkar replied:
We could leave it like this if needed, I was just pondering this thing aloud.

pfalcon replied:
Switched to 00-00-5E-00-53-FF.

jukkar reviewed (changes requested):
Couple of minor changes could be done.

I also had some work for echo-loopback sample app in https://github.com/jukkar/zephyr/pull/2 but that one did not create an interface. So this looks actually a better approach. I also had big problems in that echo-loopback app so it has not progressed lately.
